templates/support/URLSession+Api.swift.hbs

```diff
@@ -9,22 +9,46 @@ extension Foundation.URLSession {
     static var apiSession = Foundation.URLSession(configuration: .default)
 
     @available(*, renamed: "handleApiRequest(_:)")
-    static func handleApiRequest(_ request: Foundation.URLRequest, completion: @escaping ((_ response: Foundation.HTTPURLResponse?, _ data: Data?, _ error: Error?) -> Void)) {
+    static func handleApiRequest(_ request: Foundation.URLRequest, retryConfiguration: RetryConfiguration?, completion: @escaping ((_ response: Foundation.HTTPURLResponse?, _ data: Data?, _ error: Error?) -> Void)) {
         Task {
             do {
-                let result = try await handleApiRequest(request)
+                let result = try await handleApiRequest(request, retryConfiguration: retryConfiguration)
                 completion(result.response, result.data, nil)
             } catch {
                 completion(nil, nil, error)
             }
         }
     }
     
-    static func handleApiRequest(_ request: Foundation.URLRequest) async throws -> (response: HTTPURLResponse, data: Data) {
+    static func handleApiRequest(_ request: Foundation.URLRequest, attemptsTried: Int = 0, retryConfiguration: RetryConfiguration?) async throws -> URLSessionResponse {
         let (data, response) = try await apiSession.data(for: request)
         guard let httpResponse = response as? HTTPURLResponse else {
             throw APIError.unexpectedResponse(response, data: data)
         }
-        return (response: httpResponse, data: data)
+
+        let nanosecondsPerSecond: Double = 1_000_000_000
+
+        let result = URLSessionResponse(response: httpResponse, data: data)
+        guard let retryConfiguration = retryConfiguration else {
+            return result
+        }
+
+        guard attemptsTried < retryConfiguration.maxAttempts else {
+            return result
```

Comment:
Perhaps this should throw an error here to say that the number of retries was exhausted? People aren't going to expect to handle a 429 response in their own apps so that would be nicer than getting an unexpected response.

Reply:
Makes sense, I've changed it to be so